File: backend/src/compliance_platform/api/assessments.py
"""Assessment endpoints. Thin HTTP boundary only, per api/README.md:
parse the request, call the service, return the response. No
state-machine or evidence-linking logic belongs in this file — see
services/assessment_service.py.

Exception-to-HTTP-status mapping is centralized in
api/error_handlers.py (Sprint 9 refactor, ADR-0015): every custom
domain exception listed there maps to the same status code everywhere
it's raised, so endpoints below simply let it propagate rather than
catching it themselves. The one exception is bare ValueError
(review_evidence's missing corrected_practice_reference case), which is
deliberately NOT handled globally — it's too generic a type to
intercept app-wide — so it's still caught locally, right there.
"""

from __future__ import annotations

import re

from fastapi import APIRouter, Depends, HTTPException, Query, Response
from pydantic import BaseModel

from compliance_platform.api.dependencies import get_assessment_service
from compliance_platform.core.identity import get_actor
from compliance_platform.models.assessment import (
    Assessment,
    AssessmentStatus,
    AssessmentStatusChange,
    EvidenceLink,
    EvidenceRequest,
    EvidenceReviewStatus,
    EvidenceSource,
    PracticeFinding,
    PracticeFindingChange,
    PracticeFindingStatus,
    SanitizationApproval,
)
from compliance_platform.models.chat import ChatResponse
from compliance_platform.models.report import DashboardReport, EvidenceQueueSummary
from compliance_platform.models.sanitization import SanitizationPreview
from compliance_platform.models.schemas import (
    DocumentSummary,
    FinalizationReadiness,
    SealVerification,
)
from compliance_platform.services.assessment_service import AssessmentService

router = APIRouter(prefix="/assessments", tags=["assessments"])


class CreateAssessmentRequest(BaseModel):
    name: str
    framework_name: str
    # Explicit pin to a SPECIFIC framework version, if the registry has
    # more than one loaded for framework_name (Sprint 18, ADR-0053).
    # None (the default) resolves to whatever's currently latest -- the
    # same behavior this endpoint always had before this field existed.
    framework_version: str | None = None
    # Which client this assessment belongs to (Sprint 22, ADR-0063). May
    # be omitted only while exactly one organisation exists; with two or
    # more, omitting it is a 400 rather than a guess.
    organization_id: str | None = None


class StatusTransitionRequest(BaseModel):
    status: AssessmentStatus
    note: str | None = None


class LinkEvidenceRequest(BaseModel):
    document_id: str
    practice_reference: str
    chunk_id: str | None = None
    note: str | None = None
    source: EvidenceSource = EvidenceSource.MANUAL


class ReviewEvidenceRequest(BaseModel):
    decision: EvidenceReviewStatus
    corrected_practice_reference: str | None = None
    note: str | None = None


class ChatQuestionRequest(BaseModel):
    question: str


class SetPracticeFindingRequest(BaseModel):
    status: PracticeFindingStatus
    rationale: str


class RequestMoreEvidenceRequest(BaseModel):
    note: str
    # Optional since ADR-0061: the authenticated identity is preferred
    # and this is ignored whenever one is present. Kept for a direct,
    # unproxied API caller that has no other way to say who it is --
    # removing it outright would break those callers to no benefit,
    # since the server already refuses to trust it when it matters.
    requested_by: str | None = None


class ResolveEvidenceRequestRequest(BaseModel):
    resolved_by: str | None = None


@router.post("", response_model=Assessment)
def create_assessment(
    request: CreateAssessmentRequest,
    service: AssessmentService = Depends(get_assessment_service),
) -> Assessment:
    return service.create_assessment(
        name=request.name,
        framework_name=request.framework_name,
        framework_version=request.framework_version,
        organization_id=request.organization_id,
    )


@router.get("", response_model=list[Assessment])
def list_assessments(
    organization_id: str | None = None,
    service: AssessmentService = Depends(get_assessment_service),
) -> list[Assessment]:
    """Scoped to one organisation (ADR-0063). Omitting it is allowed only
    while exactly one exists, so a single-organisation deployment does
    not have to name it and a multi-client one cannot forget to."""
    return service.list_assessments(organization_id)


@router.get("/{assessment_id}", response_model=Assessment)
def get_assessment(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> Assessment:
    return service.get_assessment(assessment_id)


@router.post("/{assessment_id}/status", response_model=Assessment)
def transition_status(
    assessment_id: str,
    request: StatusTransitionRequest,
    service: AssessmentService = Depends(get_assessment_service),
    actor: str = Depends(get_actor),
) -> Assessment:
    return service.transition_status(
        assessment_id, request.status, note=request.note, actor=actor
    )


@router.get("/{assessment_id}/finalization-readiness", response_model=FinalizationReadiness)
def get_finalization_readiness(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> FinalizationReadiness:
    """Whether this assessment may be finalized, and what blocks it
    (ADR-0058).

    Blockers are machine-readable categories with the affected ids, so
    the UI renders a checklist and disables its button without parsing
    English. The same function backs the server-side gate in
    transition_status — this endpoint reports the rule, it does not
    define a second one.
    """
    return service.finalization_readiness(assessment_id)


@router.get("/{assessment_id}/verify", response_model=SealVerification)
def verify_seal(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> SealVerification:
    """Check a finalized assessment against the seal written when it was
    finalized (R-12).

    Answers the question an auditor asks about a compliance record:
    not "can your software edit this?" but "can you show nothing did?"
    The stored and recomputed digests are both returned, so anyone
    holding an exported report can compare its printed seal against
    this without trusting the verdict field.

    Deliberately a plain 200 for every outcome, including ALTERED. A
    non-2xx would make a detected alteration look like a failure to
    answer, and this endpoint answering successfully is exactly what a
    detected alteration IS.
    """
    return service.verify_finalization_seal(assessment_id)


class AttachDocumentRequest(BaseModel):
    document_id: str


@router.get("/{assessment_id}/documents", response_model=list[DocumentSummary])
def list_assessment_documents(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> list[DocumentSummary]:
    """The documents attached to this assessment (ADR-0062).

    What the evidence chooser should offer. `GET /documents` still lists
    every document on the instance and is what the attach flow browses;
    this is the scoped view.
    """
    return service.documents_for_assessment(assessment_id)


@router.post("/{assessment_id}/documents", response_model=DocumentSummary)
def attach_document(
    assessment_id: str,
    request: AttachDocumentRequest,
    service: AssessmentService = Depends(get_assessment_service),
    actor: str = Depends(get_actor),
) -> DocumentSummary:
    return service.attach_document(assessment_id, request.document_id, actor=actor)


@router.delete("/{assessment_id}/documents/{document_id}", status_code=204)
def detach_document(
    assessment_id: str,
    document_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> Response:
    """Remove a document from this assessment. Refused while evidence
    links still cite it — see AssessmentService.detach_document."""
    service.detach_document(assessment_id, document_id)
    return Response(status_code=204)


@router.get("/{assessment_id}/status-history", response_model=list[AssessmentStatusChange])
def get_status_history(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> list[AssessmentStatusChange]:
    return service.status_history(assessment_id)


@router.post("/{assessment_id}/evidence", response_model=EvidenceLink)
def link_evidence(
    assessment_id: str,
    request: LinkEvidenceRequest,
    service: AssessmentService = Depends(get_assessment_service),
    actor: str = Depends(get_actor),
) -> EvidenceLink:
    return service.link_evidence(
        actor=actor,
        assessment_id=assessment_id,
        document_id=request.document_id,
        practice_reference=request.practice_reference,
        chunk_id=request.chunk_id,
        note=request.note,
        source=request.source,
    )


@router.get("/{assessment_id}/evidence", response_model=list[EvidenceLink])
def list_evidence(
    assessment_id: str,
    review_status: EvidenceReviewStatus | None = Query(
        default=None, description="Only links in this review state."
    ),
    domain: str | None = Query(
        default=None,
        description=(
            "Only links whose practice belongs to this domain short code, resolved "
            "against the assessment's pinned framework version. An unknown code "
            "returns an empty list rather than everything."
        ),
    ),
    min_confidence: float | None = Query(
        default=None,
        ge=0.0,
        le=1.0,
        description=(
            "Only AI-proposed links at or above this retrieval similarity. NOT a "
            "calibrated probability (ADR-0011) -- see the confidence field. Manual "
            "links, which carry no confidence, are excluded rather than counted as 0."
        ),
    ),
    max_confidence: float | None = Query(default=None, ge=0.0, le=1.0),
    service: AssessmentService = Depends(get_assessment_service),
) -> list[EvidenceLink]:
    """This assessment's evidence links, optionally narrowed (ADR-0065).

    Every parameter is a view filter. None of them changes a record, and
    the response shape is unchanged from the unfiltered call, so an
    existing caller that passes nothing sees exactly what it saw before.
    """
    return service.evidence_for_assessment(
        assessment_id,
        review_status=review_status,
        domain=domain,
        min_confidence=min_confidence,
        max_confidence=max_confidence,
    )


@router.get("/{assessment_id}/evidence/summary", response_model=EvidenceQueueSummary)
def get_evidence_summary(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> EvidenceQueueSummary:
    """What the whole queue holds, so a filtered view can say what it is
    a subset of (ADR-0065). Deliberately takes no filter parameters --
    a total that moves with the filter answers nothing.
    """
    return service.evidence_queue_summary(assessment_id)


@router.get("/{assessment_id}/score", response_model=dict[str, float])
def get_scores(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> dict[str, float]:
    """Per-domain score. Meaning depends on the assessment's framework —
    check GET /frameworks/{name}'s scoring_model field: "cumulative_mil"
    means a whole-number MIL 0-3 (C2M2); "coverage" means a 0.0-1.0
    fraction of subcategories with evidence (NIST CSF 2.0, which has no
    native maturity concept — see ADR-0010). A domain not yet
    transcribed into framework_mapping/ always reports 0/0.0, not an
    error — see Domain.practices_populated.
    """
    return service.compute_scores(assessment_id)


@router.get("/{assessment_id}/dashboard", response_model=DashboardReport)
def get_dashboard(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> DashboardReport:
    """Executive dashboard (Sprint 6): situation/complication/resolution
    view of this assessment — see services/report_service.py and
    ADR-0012. No LLM narrative generation; every figure is computed
    directly from real evidence links and the framework's structured
    schema.
    """
    return service.build_dashboard(assessment_id)


_SLUG_INVALID_CHARS = re.compile(r"[^A-Za-z0-9._-]+")


def _slugify_filename(name: str) -> str:
    slug = _SLUG_INVALID_CHARS.sub("-", name).strip("-")
    return slug or "assessment"


@router.get("/{assessment_id}/report/pdf")
def get_dashboard_pdf(
    assessment_id: str,
    sanitized: bool = False,
    service: AssessmentService = Depends(get_assessment_service),
) -> Response:
    """PDF rendering of the same dashboard GET /dashboard returns
    (Sprint 7) — see services/export_service.py and ADR-0013. Same
    error mapping as the dashboard endpoint, since both are built from
    the same DashboardReport. sanitized=true (ADR-0032) requires a
    current, matching PracticeFinding/evidence-state approval recorded
    via POST .../sanitization/approve first — see
    services/assessment_service.py.
    """
    assessment = service.get_assessment(assessment_id)
    pdf_bytes = service.generate_dashboard_pdf(assessment_id, sanitized=sanitized)
    suffix = "_sanitized" if sanitized else ""
    filename = f"{_slugify_filename(assessment.name)}_dashboard{suffix}.pdf"
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.get("/{assessment_id}/report/xlsx")
def get_dashboard_xlsx(
    assessment_id: str,
    sanitized: bool = False,
    service: AssessmentService = Depends(get_assessment_service),
) -> Response:
    """XLSX rendering of the same dashboard GET /dashboard returns
    (Sprint 7) — see services/export_service.py and ADR-0013. See
    get_dashboard_pdf's docstring for the sanitized=true behavior.
    """
    assessment = service.get_assessment(assessment_id)
    xlsx_bytes = service.generate_dashboard_xlsx(assessment_id, sanitized=sanitized)
    suffix = "_sanitized" if sanitized else ""
    filename = f"{_slugify_filename(assessment.name)}_dashboard{suffix}.xlsx"
    return Response(
        content=xlsx_bytes,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


class SanitizationPreviewRequest(BaseModel):
    custom_terms: list[str] = []


class ApproveSanitizationRequest(BaseModel):
    custom_terms: list[str] = []
    approved_by: str


@router.post("/{assessment_id}/sanitization/preview", response_model=SanitizationPreview)
def preview_sanitization(
    assessment_id: str,
    request: SanitizationPreviewRequest,
    service: AssessmentService = Depends(get_assessment_service),
) -> SanitizationPreview:
    """Read-only preview/diff of what a sanitized export would redact
    or pseudonymize (ADR-0032) — never persisted, never itself
    authorizes an export. custom_terms is the reviewer-supplied list of
    organization-specific identifiers (names, facility/vendor/customer/
    employee identifiers) to pseudonymize; see
    services/sanitization_service.py for why those categories can't be
    detected automatically without fabricating an NER capability this
    project hasn't evaluated.
    """
    return service.preview_sanitization(assessment_id, request.custom_terms)


@router.post("/{assessment_id}/sanitization/approve", response_model=SanitizationApproval)
def approve_sanitization(
    assessment_id: str,
    request: ApproveSanitizationRequest,
    service: AssessmentService = Depends(get_assessment_service),
) -> SanitizationApproval:
    """Records explicit human approval of one specific sanitized report
    (ADR-0032's "never silently publish an AI-sanitized report" rule).
    Required before GET .../report/pdf|xlsx?sanitized=true will
    succeed; a later change to the underlying report invalidates this
    approval automatically (see SanitizationApprovalStaleError).
    """
    return service.approve_sanitization(
        assessment_id, request.custom_terms, request.approved_by
    )


@router.post("/{assessment_id}/evidence/{evidence_link_id}/review", response_model=EvidenceLink)
def review_evidence(
    assessment_id: str,
    evidence_link_id: str,
    request: ReviewEvidenceRequest,
    service: AssessmentService = Depends(get_assessment_service),
    actor: str = Depends(get_actor),
) -> EvidenceLink:
    """Applies a human accept/edit/reject decision to a pending evidence
    link — see services/assessment_service.py.review_evidence and the
    assessment-generation skill's human-in-the-loop invariant. Only
    ValueError (a missing corrected_practice_reference on an "edited"
    decision) is caught here — every other exception this can raise is
    handled globally, see api/error_handlers.py.
    """
    try:
        return service.review_evidence(
            assessment_id=assessment_id,
            evidence_link_id=evidence_link_id,
            decision=request.decision,
            corrected_practice_reference=request.corrected_practice_reference,
            note=request.note,
            actor=actor,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.put(
    "/{assessment_id}/practice-findings/{practice_reference}", response_model=PracticeFinding
)
def set_practice_finding(
    assessment_id: str,
    practice_reference: str,
    request: SetPracticeFindingRequest,
    service: AssessmentService = Depends(get_assessment_service),
    actor: str = Depends(get_actor),
) -> PracticeFinding:
    """Records a reviewer's explicit compliance judgment for one
    practice — SATISFIED, PARTIALLY_SATISFIED, NOT_SATISFIED,
    INSUFFICIENT_EVIDENCE, or NOT_APPLICABLE (ADR-0030). PUT, not POST:
    idempotent upsert of the single current finding for this
    (assessment, practice) pair, not a growing list of independent
    records — see services/assessment_service.py.set_practice_finding
    and PracticeFindingChange for the append-only history this still
    preserves underneath.
    """
    return service.set_practice_finding(
        assessment_id=assessment_id,
        practice_reference=practice_reference,
        status=request.status,
        rationale=request.rationale,
        actor=actor,
    )


@router.get("/{assessment_id}/practice-findings", response_model=list[PracticeFinding])
def list_practice_findings(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> list[PracticeFinding]:
    return service.practice_findings_for_assessment(assessment_id)


@router.get(
    "/{assessment_id}/practice-findings/{practice_reference}/history",
    response_model=list[PracticeFindingChange],
)
def get_practice_finding_history(
    assessment_id: str,
    practice_reference: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> list[PracticeFindingChange]:
    return service.practice_finding_history(assessment_id, practice_reference)


@router.post(
    "/{assessment_id}/practice-findings/{practice_reference}/evidence-requests",
    response_model=EvidenceRequest,
)
def request_more_evidence(
    assessment_id: str,
    practice_reference: str,
    request: RequestMoreEvidenceRequest,
    service: AssessmentService = Depends(get_assessment_service),
    actor: str = Depends(get_actor),
) -> EvidenceRequest:
    """Records a reviewer's explicit request that someone go find and
    upload more evidence for this practice (ADR-0043) — a workflow
    action distinct from set_practice_finding's compliance judgment;
    the two can coexist for the same practice. POST, not PUT: each call
    creates a new request, not an idempotent upsert of a single current
    one — multiple open requests can exist for the same practice.
    """
    return service.request_more_evidence(
        assessment_id=assessment_id,
        practice_reference=practice_reference,
        note=request.note,
        requested_by=request.requested_by,
        actor=actor,
    )


@router.get("/{assessment_id}/evidence-requests", response_model=list[EvidenceRequest])
def list_evidence_requests(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
) -> list[EvidenceRequest]:
    return service.evidence_requests_for_assessment(assessment_id)


@router.post(
    "/{assessment_id}/evidence-requests/{request_id}/resolve", response_model=EvidenceRequest
)
def resolve_evidence_request(
    assessment_id: str,
    request_id: str,
    request: ResolveEvidenceRequestRequest,
    service: AssessmentService = Depends(get_assessment_service),
    actor: str = Depends(get_actor),
) -> EvidenceRequest:
    """Marks an evidence request resolved. Always explicit -- never
    inferred from a new evidence link being added, since linking
    evidence doesn't guarantee it actually addresses what was
    requested.
    """
    return service.resolve_evidence_request(
        assessment_id=assessment_id,
        request_id=request_id,
        resolved_by=request.resolved_by,
        actor=actor,
    )


@router.post("/{assessment_id}/propose-mappings", response_model=list[EvidenceLink])
def propose_mappings(
    assessment_id: str,
    service: AssessmentService = Depends(get_assessment_service),
    actor: str = Depends(get_actor),
) -> list[EvidenceLink]:
    """Runs the retrieval-based mapping engine (services/mapping_service.py,
    ADR-0011) and persists any resulting proposals as AI-proposed,
    pending-review evidence links — always over documents already
    associated with this assessment, never the whole vector store.
    """
    return service.propose_mappings(assessment_id, actor=actor)


@router.post("/{assessment_id}/chat", response_model=ChatResponse)
def chat_with_assessment(
    assessment_id: str,
    request: ChatQuestionRequest,
    service: AssessmentService = Depends(get_assessment_service),
) -> ChatResponse:
    """Retrieval-only Q&A over this assessment's reviewed evidence
    (Sprint 8) — see services/chat_service.py and ADR-0014. Returns
    ranked, cited evidence chunks; nothing here is model-generated, so
    an empty result list (no reviewed evidence, or nothing above the
    similarity threshold) is a valid 200 response, not an error.
    """
    return service.answer_question(assessment_id, request.question)

File: frontend/src/api/assessments.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { API_BASE_URL, apiClient } from './client'
import type {
  Assessment,
  AssessmentStatusChange,
  ChatResponse,
  CreateAssessmentRequest,
  DashboardReport,
  EvidenceLink,
  EvidenceQueueSummary,
  EvidenceRequest,
  EvidenceReviewStatus,
  DocumentSummary,
  FinalizationReadiness,
  SealVerification,
  LinkEvidenceRequest,
  PracticeFinding,
  PracticeFindingChange,
  PracticeFindingStatus,
  ReviewEvidenceRequest,
  SanitizationApproval,
  SanitizationPreview,
  StatusTransitionRequest,
} from './types'

const keys = {
  // The organisation is part of the key, not just the URL (ADR-0063).
  // Without it, switching client would serve the previous one's
  // assessments out of the cache -- a stale list is a nuisance
  // everywhere else in this app and a confidentiality failure here.
  assessments: (organizationId: string | undefined) =>
    ['assessments', 'list', organizationId ?? ''] as const,
  assessment: (id: string) => ['assessments', id] as const,
  statusHistory: (id: string) => ['assessments', id, 'status-history'] as const,
  // The filter is part of the key. Without it, changing the filter
  // would serve the previous filter's rows out of the cache and the
  // control would look broken; with it, an invalidation after a
  // review still refreshes every filtered view of the same queue,
  // because they all share this prefix.
  evidence: (id: string, filters?: EvidenceFilters) =>
    ['assessments', id, 'evidence', filters ?? {}] as const,
  evidenceSummary: (id: string) => ['assessments', id, 'evidence', 'summary'] as const,
  // Every filtered list AND the summary sit under this prefix, so a
  // review decision invalidates all of them with one call. Passing
  // keys.evidence(id) here instead would only match the unfiltered
  // list, leaving a reviewer working inside a filter looking at a row
  // they had just decided on.
  evidenceAll: (id: string) => ['assessments', id, 'evidence'] as const,
  dashboard: (id: string) => ['assessments', id, 'dashboard'] as const,
  practiceFindings: (id: string) => ['assessments', id, 'practice-findings'] as const,
  practiceFindingHistory: (id: string, practiceReference: string) =>
    ['assessments', id, 'practice-findings', practiceReference, 'history'] as const,
  evidenceRequests: (id: string) => ['assessments', id, 'evidence-requests'] as const,
  finalizationReadiness: (id: string) =>
    ['assessments', id, 'finalization-readiness'] as const,
  sealVerification: (id: string) => ['assessments', id, 'seal-verification'] as const,
  assessmentDocuments: (id: string) => ['assessments', id, 'documents'] as const,
}

export function useAssessments(organizationId: string | undefined) {
  return useQuery({
    queryKey: keys.assessments(organizationId),
    queryFn: () =>
      apiClient.get<Assessment[]>(
        `/assessments?organization_id=${encodeURIComponent(organizationId ?? '')}`,
      ),
    // Nothing is fetched before the organisation is known: an unscoped
    // request would either 400 (two or more organisations) or quietly
    // answer for the only one, and the second is the habit worth not
    // forming.
    enabled: !!organizationId,
  })
}

export function useAssessment(id: string | undefined) {
  return useQuery({
    queryKey: keys.assessment(id ?? ''),
    queryFn: () => apiClient.get<Assessment>(`/assessments/${id}`),
    enabled: !!id,
  })
}

export function useCreateAssessment(organizationId: string | undefined) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (body: CreateAssessmentRequest) =>
      apiClient.post<Assessment>('/assessments', { ...body, organization_id: organizationId }),
    onSuccess: () =>
      queryClient.invalidateQueries({ queryKey: keys.assessments(organizationId) }),
  })
}

export function useStatusHistory(assessmentId: string | undefined) {
  return useQuery({
    queryKey: keys.statusHistory(assessmentId ?? ''),
    queryFn: () =>
      apiClient.get<AssessmentStatusChange[]>(`/assessments/${assessmentId}/status-history`),
    enabled: !!assessmentId,
  })
}

export function useTransitionStatus(assessmentId: string) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (body: StatusTransitionRequest) =>
      apiClient.post<Assessment>(`/assessments/${assessmentId}/status`, body),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.assessment(assessmentId) })
      queryClient.invalidateQueries({ queryKey: keys.statusHistory(assessmentId) })
      // By prefix, without an organisation: a status change should
      // refresh whichever list is on screen, and this hook does not know
      // which organisation that is.
      queryClient.invalidateQueries({ queryKey: ['assessments', 'list'] })
    },
  })
}

export type EvidenceFilters = {
  review_status?: EvidenceReviewStatus
  domain?: string
  min_confidence?: number
  max_confidence?: number
}

function evidenceQueryString(filters: EvidenceFilters | undefined): string {
  if (!filters) return ''
  const params = new URLSearchParams()
  for (const [key, value] of Object.entries(filters)) {
    // An empty string is what a cleared <select> gives back, and it is
    // not a filter — sending it would ask the server for links whose
    // domain is "".
    if (value === undefined || value === '') continue
    params.set(key, String(value))
  }
  const query = params.toString()
  return query ? `?${query}` : ''
}

export function useEvidenceLinks(
  assessmentId: string | undefined,
  filters?: EvidenceFilters,
) {
  return useQuery({
    queryKey: keys.evidence(assessmentId ?? '', filters),
    queryFn: () =>
      apiClient.get<EvidenceLink[]>(
        `/assessments/${assessmentId}/evidence${evidenceQueryString(filters)}`,
      ),
    enabled: !!assessmentId,
  })
}

/**
 * Counts over the whole queue, never the filtered view (ADR-0065).
 *
 * Kept a separate request rather than folded into the list response so
 * that the totals a reviewer reads cannot move when they change a
 * filter — a "23 of 412" that becomes "23 of 23" answers nothing.
 */
export function useEvidenceSummary(assessmentId: string | undefined) {
  return useQuery({
    queryKey: keys.evidenceSummary(assessmentId ?? ''),
    queryFn: () =>
      apiClient.get<EvidenceQueueSummary>(`/assessments/${assessmentId}/evidence/summary`),
    enabled: !!assessmentId,
  })
}

export function useLinkEvidence(assessmentId: string) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (body: LinkEvidenceRequest) =>
      apiClient.post<EvidenceLink>(`/assessments/${assessmentId}/evidence`, body),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: keys.evidenceAll(assessmentId) }),
  })
}

export function useProposeMappings(assessmentId: string) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: () => apiClient.post<EvidenceLink[]>(`/assessments/${assessmentId}/propose-mappings`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.evidenceAll(assessmentId) })
      // New AI proposals are unreviewed by definition, so proposing
      // mappings BLOCKS finalization until they are reviewed.
      queryClient.invalidateQueries({ queryKey: keys.finalizationReadiness(assessmentId) })
    },
  })
}

export function useReviewEvidence(assessmentId: string) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ evidenceLinkId, body }: { evidenceLinkId: string; body: ReviewEvidenceRequest }) =>
      apiClient.post<EvidenceLink>(
        `/assessments/${assessmentId}/evidence/${evidenceLinkId}/review`,
        body,
      ),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.evidenceAll(assessmentId) })
      // review outcomes feed directly into score/dashboard
      queryClient.invalidateQueries({ queryKey: keys.dashboard(assessmentId) })
      // ...and into whether the assessment may be finalized (ADR-0058):
      // an unreviewed proposal is a blocker, so accepting the last one
      // must re-enable Finalize without a page reload.
      queryClient.invalidateQueries({ queryKey: keys.finalizationReadiness(assessmentId) })
    },
  })
}

/**
 * Whether this assessment may be finalized, and what blocks it
 * (ADR-0058).
 *
 * The same rule is enforced server-side in transition_status — this
 * exists so the reviewer can see and fix the blockers rather than
 * discovering them as a 409 after pressing a button. It is a usability
 * affordance, not the integrity boundary.
 */
export function useFinalizationReadiness(assessmentId: string | undefined) {
  return useQuery({
    queryKey: keys.finalizationReadiness(assessmentId ?? ''),
    queryFn: () =>
      apiClient.get<FinalizationReadiness>(
        `/assessments/${assessmentId}/finalization-readiness`,
      ),
    enabled: !!assessmentId,
  })
}

/**
 * Check a finalized assessment against the seal written when it was
 * finalized (ADR-0060).
 *
 * Deliberately NOT run on render. Verification is a question somebody
 * asks — "is this record still the record?" — and an answer that
 * appears before anyone asked reads as decoration. It also costs a full
 * re-read and re-hash of the assessment, which is not something to do
 * every time a tab mounts.
 */
export function useVerifySeal(assessmentId: string | undefined) {
  return useQuery({
    queryKey: keys.sealVerification(assessmentId ?? ''),
    queryFn: () => apiClient.get<SealVerification>(`/assessments/${assessmentId}/verify`),
    enabled: false,
    // A verdict has a moment attached to it. Keeping a stale one around
    // would let the panel say "verified" about a check run before an
    // edit that has happened since.
    gcTime: 0,
  })
}

/**
 * The documents attached to this assessment (ADR-0062).
 *
 * What the evidence chooser offers. It used to call `useDocuments()`,
 * which lists every document on the instance — so a reviewer picking
 * evidence for one organisation's assessment was shown another
 * organisation's policies, and could link them.
 */
export function useAssessmentDocuments(assessmentId: string | undefined) {
  return useQuery({
    queryKey: keys.assessmentDocuments(assessmentId ?? ''),
    queryFn: () => apiClient.get<DocumentSummary[]>(`/assessments/${assessmentId}/documents`),
    enabled: !!assessmentId,
  })
}

export function useAttachDocument(assessmentId: string) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (documentId: string) =>
      apiClient.post<DocumentSummary>(`/assessments/${assessmentId}/documents`, {
        document_id: documentId,
      }),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.assessmentDocuments(assessmentId) })
    },
  })
}

export function useDetachDocument(assessmentId: string) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (documentId: string) =>
      apiClient.del<void>(`/assessments/${assessmentId}/documents/${documentId}`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.assessmentDocuments(assessmentId) })
    },
  })
}

export function useDashboard(assessmentId: string | undefined) {
  return useQuery({
    queryKey: keys.dashboard(assessmentId ?? ''),
    queryFn: () => apiClient.get<DashboardReport>(`/assessments/${assessmentId}/dashboard`),
    enabled: !!assessmentId,
  })
}

export function useChat(assessmentId: string) {
  return useMutation({
    mutationFn: (question: string) =>
      apiClient.post<ChatResponse>(`/assessments/${assessmentId}/chat`, { question }),
  })
}

export function reportUrl(
  assessmentId: string,
  format: 'pdf' | 'xlsx',
  sanitized = false,
): string {
  const suffix = sanitized ? '?sanitized=true' : ''
  return `${API_BASE_URL}/assessments/${assessmentId}/report/${format}${suffix}`
}

// --- Practice findings (ADR-0030) ---

export function usePracticeFindings(assessmentId: string | undefined) {
  return useQuery({
    queryKey: keys.practiceFindings(assessmentId ?? ''),
    queryFn: () => apiClient.get<PracticeFinding[]>(`/assessments/${assessmentId}/practice-findings`),
    enabled: !!assessmentId,
  })
}

export function usePracticeFindingHistory(assessmentId: string, practiceReference: string | null) {
  return useQuery({
    queryKey: keys.practiceFindingHistory(assessmentId, practiceReference ?? ''),
    queryFn: () =>
      apiClient.get<PracticeFindingChange[]>(
        `/assessments/${assessmentId}/practice-findings/${practiceReference}/history`,
      ),
    enabled: !!practiceReference,
  })
}

export function useSetPracticeFinding(assessmentId: string) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({
      practiceReference,
      status,
      rationale,
    }: {
      practiceReference: string
      status: PracticeFindingStatus
      rationale: string
    }) =>
      apiClient.put<PracticeFinding>(
        `/assessments/${assessmentId}/practice-findings/${practiceReference}`,
        { status, rationale },
      ),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.practiceFindings(assessmentId) })
      // findings feed directly into score/dashboard (ADR-0030)
      queryClient.invalidateQueries({ queryKey: keys.dashboard(assessmentId) })
      // ...and a SATISFIED/NOT_APPLICABLE finding without supporting
      // evidence is itself a finalization blocker (ADR-0058).
      queryClient.invalidateQueries({ queryKey: keys.finalizationReadiness(assessmentId) })
    },
  })
}

// --- Evidence requests (ADR-0043) ---

export function useEvidenceRequests(assessmentId: string | undefined) {
  return useQuery({
    queryKey: keys.evidenceRequests(assessmentId ?? ''),
    queryFn: () =>
      apiClient.get<EvidenceRequest[]>(`/assessments/${assessmentId}/evidence-requests`),
    enabled: !!assessmentId,
  })
}

export function useRequestMoreEvidence(assessmentId: string) {
  const queryClient = useQueryClient()
  return useMutation({
    // No requested_by: the server attributes this to the authenticated
    // identity and ignores anything the client claims (ADR-0061).
    mutationFn: ({ practiceReference, note }: { practiceReference: string; note: string }) =>
      apiClient.post<EvidenceRequest>(
        `/assessments/${assessmentId}/practice-findings/${practiceReference}/evidence-requests`,
        { note },
      ),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.evidenceRequests(assessmentId) })
      queryClient.invalidateQueries({ queryKey: keys.finalizationReadiness(assessmentId) })
    },
  })
}

export function useResolveEvidenceRequest(assessmentId: string) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ requestId }: { requestId: string }) =>
      apiClient.post<EvidenceRequest>(
        `/assessments/${assessmentId}/evidence-requests/${requestId}/resolve`,
        {},
      ),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: keys.evidenceRequests(assessmentId) })
      queryClient.invalidateQueries({ queryKey: keys.finalizationReadiness(assessmentId) })
    },
  })
}

// --- Sanitization (ADR-0032) ---

export function usePreviewSanitization(assessmentId: string) {
  return useMutation({
    mutationFn: (customTerms: string[]) =>
      apiClient.post<SanitizationPreview>(`/assessments/${assessmentId}/sanitization/preview`, {
        custom_terms: customTerms,
      }),
  })
}

export function useApproveSanitization(assessmentId: string) {
  return useMutation({
    mutationFn: ({ customTerms, approvedBy }: { customTerms: string[]; approvedBy: string }) =>
      apiClient.post<SanitizationApproval>(`/assessments/${assessmentId}/sanitization/approve`, {
        custom_terms: customTerms,
        approved_by: approvedBy,
      }),
  })
}
